File: Chess-Challenge/src/NEAT Bot/InnovationHandler.cs
using System;
using System.Collections.Generic;

namespace Chess_Challenge.NEAT_Bot; 

/**
 * This class should be created as a singleton and passed into each Organism
 */
public class InnovationHandler {
    private int _curOrganismId = 0;
    private int _curNodeId = 0;
    private int _curConnectionId = 0;
    public Dictionary<Tuple<int, int>, int> NodeIdMapping = new();
    public Dictionary<Tuple<int, int>, int> ConnectionIdMapping = new();

    public InnovationHandler(int numInputs, int numOutputs) {
        _curNodeId = numInputs + numOutputs;
    }

    public int GetNextOrganismId() {
        var curId = _curOrganismId;
        _curOrganismId++;
        return curId;
    }
    
    public int AssignNodeId(Tuple<int, int> source) {
        int assignedId;
        if (NodeIdMapping.TryGetValue(source, out var value)) {
            // If the source connection is already in the map, use the same innovation number
            assignedId = value;
        } else {
            // If not, get a new innovation number and add it to the map
            assignedId = _curNodeId;
            NodeIdMapping[source] = assignedId;
            _curNodeId++;
        }
        return assignedId;
    }
    
    public int AssignConnectionId(Tuple<int, int> connection) {
        if (ConnectionIdMapping.TryGetValue(connection, out var value)) {
            // If the connection is already in the map, return its ID
            return value;
        } else {
            // If not, create a new ID and add it to the map
            int newId = _curConnectionId;
            ConnectionIdMapping[connection] = newId;
            _curConnectionId++;
            return newId;
        }
    }
}

File: python_neat/src/innovation_handler.py
from constants import Constants


class InnovationHandler:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(InnovationHandler, cls).__new__(cls)
            cls._curOrganismId = 0
            cls._curNodeId = Constants.inputs_count + Constants.outputs_count
            cls._curConnectionId = 0
            cls.node_id_mapping = {}
            cls.connection_id_mapping = {}
        return cls._instance

    def get_next_organism_id(self):
        cur_id = self._curOrganismId
        self._curOrganismId += 1
        return cur_id

    def assign_node_id(self, source):
        if source in self.node_id_mapping:
            assigned_id = self.node_id_mapping[source]
        else:
            assigned_id = self._curNodeId
            self._curNodeId += 1

            if source is not None:
                self.node_id_mapping[source] = assigned_id

        return assigned_id

    def assign_connection_id(self, connection):
        if connection in self.connection_id_mapping:
            return self.connection_id_mapping[connection]
        else:
            new_id = self._curConnectionId
            self.connection_id_mapping[connection] = new_id
            self._curConnectionId += 1
            return new_id
